soroban-sdk/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,294 @@
+/// let res = bls12_381.g1_add(&zero, &one);
+/// assert_eq!(res, one);
+/// ```
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct G1Affine(BytesN<96>);
+
+/// # `G2Affine` is a point in the G2 group (subgroup defined over the quadratic
+/// extension field `Fq2`) of the BLS12-381 elliptic curve
+///
+/// # Serialization:
+/// - The 192 bytes represent the **uncompressed encoding** of a point in G2.
+///   The bytes consist of `be_bytes(X_c1) || be_bytes(X_c0) || be_bytes(Y_c1)
+///   || be_bytes(Y_c0)` (`||` is concatenation), where 'X' and 'Y' are the two
+///   coordinates, each being an extension field element `Fp2` and `c0`, `c1`
+///   are components of `Fp2` (each being `Fp`).
+/// - The most significant three bits (bits 0-3) of the first byte are reserved
+///   for encoding flags:
+///   - compression_flag (bit 0): Must always be set (1), as only uncompressed
+///     points are supported.
+///   - infinity_flag (bit 1): Set if the point is the point at infinity (zero
+///     point), in which case all other bits must be zero.
+///   - sort_flag (bit 2): Must always be unset (0).
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct G2Affine(BytesN<192>);
+
+/// # `Fp` represents an element of the base field `Fq` of the BLS12-381 elliptic
+/// curve
+///
+/// # Serialization:
+/// - The 48 bytes represent the **big-endian encoding** of an element in the
+///   field `Fp`. The value is serialized as a big-endian integer.
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Fp(BytesN<48>);
+
+/// # `Fp2` represents an element of the quadratic extension field `Fq2` of the
+/// BLS12-381 elliptic curve
+///
+/// # Serialization:
+/// - The 96 bytes represent the **big-endian encoding** of an element in the
+///   field `Fp2`. The bytes consist of `be_bytes(c1) || be_bytes(c0)` (`||` is
+///   concatenation), where `c0` and `c1` are the two `Fp` elements (the real
+///   and imaginary components).
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Fp2(BytesN<96>);
+
+impl_bytesn_repr!(G1Affine, 96);
+impl_bytesn_repr!(G2Affine, 192);
+impl_bytesn_repr!(Fp, 48);
+impl_bytesn_repr!(Fp2, 96);
+
+impl Bls12_381 {
+    pub(crate) fn new(env: &Env) -> Bls12_381 {
+        Bls12_381 { env: env.clone() }
+    }
+
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    // g1
+
+    /// Checks if a point `p` in G1 is in the correct subgroup.
+    pub fn g1_is_in_subgroup(&self, p: &G1Affine) -> bool {
+        let env = self.env();
+        let res = internal::Env::bls12_381_check_g1_is_in_subgroup(env, p.to_object())
+            .unwrap_infallible();
+        res.into()
+    }
+
+    /// Adds two points `p0` and `p1` in G1.
+    pub fn g1_add(&self, p0: &G1Affine, p1: &G1Affine) -> G1Affine {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g1_add(env, p0.to_object(), p1.to_object())
+            .unwrap_infallible();
+        unsafe { G1Affine::from_bytes(<BytesN<96>>::unchecked_new(env.clone(), bin)) }
```

Comment:
Fyi the size of BytesN can be inferred by the destination, so it doesn't need mentioning at all the call sites.
```suggestion
        unsafe { G1Affine::from_bytes(BytesN::unchecked_new(env.clone(), bin)) }
```

Reply:
You don't have to do this, but you may wish to add a G1Affine::unchecked_new. Most of the SDK types have the function.